### PR TITLE
Provided mapping from vscode command ids to internal command ids

### PR DESCRIPTION
**Pull request by JPinkney**
#### What it does

This PR introduces mappings from vscode command ids to internal theia commands ids as part of https://github.com/theia-ide/theia/issues/5113 and https://github.com/theia-ide/theia/issues/4247.

All commands listed [here](https://code.visualstudio.com/docs/getstarted/keybindings) that are not registered under vscode's command id but instead have a command id native to theia are mapped. E.g. workbench.action.files.saveAs is mapped to file.saveAs so that when an extension calls workbench.action.files.saveAs the command that actually gets executed is file.saveAs.
However, there are still a lot of commands missing that either don't have an internal theia command that is relatable or don't have an implementation at all.

Also, when commands are being registered inside of plugins, vscode command ids must be mapped to their corresponding internal command id so that when you use that keybinding it understands to use the mapped internal command id rather than the vscode command id. This is because without the mapping the keybindings thinks that it should execute editor.action.formatSelection which doesn't have a registered handler, rather than monaco.editor.action.formatSelection which does have a handler.

After looking into multiple approaches on how to solve this problem this seems to be the best approach I can find. However, there are some shortcomings to this approach: 

- Due to the fact we have to map vscode commands to internal theia commands, attempting to add in something like editor.action.formatSelection into keymaps.json will not work because keymaps.json expects that command to be monaco.editor.action.formatSelection.

- Finding the vscode command in the keybindings widget will fail because theia's internal command id would be monaco.editor.action.formatSelection rather than editor.action.formatSelection.

As far as I could find, this is the best solution that is doesn't change core or break any existing theia extensions.

#### How to test

Install **vscode-java**, **alphabotsec.vscode-eclipse-keybindings**
Then clone git@github.com:JPinkney/Theia-Extension-Test-Repo.git

**How to test execute commands:**
Set "java.referencesCodeLens.enabled": true

Open 'helloworld.java'
From the command pallete run 'open Java language server log file'
From within that java file test references
Then start the hosted instance on the cloned project and run `Test by commenting this line` from the command palette 

**How to test the keybinding mappings work:**
Try these mapped keybindings on the helloworld.java file
alt+shift+r/cmd+alt+r -> rename a variable
ctrl+alt+j/cmd+alt+j -> join lines
ctrl/cmd+shift+c -> comments the line
ctrl+h -> brings up search in side search bar
cmd+l/ctrl+l -> brings up go to line prompt

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Josh Pinkney <joshpinkney@gmail.com>

### CODE REVIEW

**Comment by JPinkney** [comment]
This is a master list of all missing vscode commands from the [vscode keybinding page](https://code.visualstudio.com/docs/getstarted/keybindings) that this PR provides.

A 1 in the **registered correctly column** means the internal command id representation is the same as vscode api's (and thus nothing needs to be done because its already correctly registered). A 0 or nothing means that there is no command with that command id.

A 1 in the **forwarded in plugin** column means that the vscode command is is mapped to an internal theia command id.


Command ID | Registered Correctly | Forwarded in Plugin (If there is a compatible command)
-- | -- | --
**Rich Languages Editing** |   | 
editor.action.triggerSuggest |   | 1
editor.action.triggerParameterHints |   | 1
editor.action.formatDocument |   | 1
editor.action.formatSelection |   | 1
editor.action.revealDefinition | 0 | 0
editor.action.showHover |   | 1
editor.action.peekDefinition | 0 | 0
editor.action.revealDefinitionAside | 0 | 0
editor.action.quickFix |   | 1
editor.action.referenceSearch.trigger |   | 1
editor.action.rename |   | 1
editor.action.inPlaceReplace.down |   | 1
editor.action.inPlaceReplace.up |   | 1
editor.action.smartSelect.expand | 0 | 0
editor.action.smartSelect.shrink |   | 1
editor.action.trimTrailingWhitespace |   | 1
workbench.action.editor.changeLanguageMode | 0 | 0
  |   |  
**Navigation** |   |  
workbench.action.showAllSymbols | 0 | 0
workbench.action.gotoLine |   | 1
workbench.action.quickOpen | 0 | 0
workbench.action.gotoSymbol |   | 1
workbench.actions.view.problems |   | 1
editor.action.marker.nextInFiles |   | 1
editor.action.marker.prevInFiles |   | 1
workbench.action.showCommands | 1 | 0
workbench.action.openPreviousRecentlyUsedEditorInGroup | 0 | 0
workbench.action.navigateBack | 0 | 0
workbench.action.quickInputBack | 0 | 0
workbench.action.navigateForward | 0 | 0
  |   |  
**Editor/Window Management** |   |  
workbench.action.newWindow | 0 |  
workbench.action.closeWindow | 0 |  
workbench.action.closeActiveEditor | 1 |  
workbench.action.closeFolder | 0 |  
workbench.action.navigateEditorGroups | 0 |  
workbench.action.splitEditor | 0 |  
workbench.action.focusFirstEditorGroup | 0 |  
workbench.action.focusSecondEditorGroup | 0 |  
workbench.action.focusThirdEditorGroup | 0 |  
workbench.action.focusPreviousGroup | 0 |  
workbench.action.focusNextGroup | 0 |  
workbench.action.moveEditorLeftInGroup | 0 |  
workbench.action.moveEditorRightInGroup | 0 |  
workbench.action.moveActiveEditorGroupLeft | 0 |  
workbench.action.moveActiveEditorGroupRight | 0 |  
workbench.action.moveEditorToNextGroup | 0 |  
workbench.action.moveEditorToPreviousGroup | 0 |  
  |   |  
**File Management** |   |  
workbench.action.files.newUntitledFile | 1 |  
workbench.action.files.openFile | 1 |  
workbench.action.files.save | 1 |  
workbench.action.files.saveAll | 1 |  
workbench.action.files.saveAs |   | 1
workbench.action.closeActiveEditor | 1 |  
workbench.action.closeOtherEditors | 1 |  
workbench.action.closeEditorsInGroup | 1 |  
workbench.action.closeEditorsInOtherGroups | 1 |  
workbench.action.closeEditorsToTheLeft | 1 |  
workbench.action.closeEditorsToTheRight | 1 |  
workbench.action.closeAllEditors | 1 |  
workbench.action.reopenClosedEditor | 0 |  
workbench.action.keepEditor | 0 |  
workbench.action.openNextRecentlyUsedEditorInGroup | 0 |  
workbench.action.openPreviousRecentlyUsedEditorInGroup | 0 |  
workbench.action.files.copyPathOfActiveFile | 0 |  
workbench.action.files.revealActiveFileInWindows | 0 |  
workbench.action.files.showOpenedFileInNewWindow | 0 |  
workbench.files.action.compareFileWith | 0 |  
  |   |  
**Display** |   |  
workbench.action.toggleFullScreen | 0 |  
workbench.action.toggleZenMode | 0 |  
workbench.action.exitZenMode | 0 |  
workbench.action.zoomIn | 0 |  
workbench.action.zoomOut | 0 |  
workbench.action.zoomReset | 0 |  
workbench.action.toggleSidebarVisibility | 0 |  
workbench.view.explorer |   | 1
workbench.view.search |   | 1
workbench.view.scm |   | 1
workbench.view.debug |   | 1
workbench.view.extensions |   | 1
workbench.action.output.toggleOutput |   | 1
workbench.action.quickOpenView | 0 |  
workbench.action.terminal.openNativeConsole |   | 1
markdown.showPreview | 0 |  
markdown.showPreviewToSide | 0 |  
workbench.action.terminal.toggleTerminal | 0 |  
  |   |  
**Search** |   |  
workbench.view.search |   | 1
workbench.action.replaceInFiles | 0 |  
toggleSearchCaseSensitive | 0 |  
toggleSearchWholeWord | 0 |  
toggleSearchRegex | 0 |  
workbench.action.search.toggleQueryDetails | 0 |  
search.action.focusNextSearchResult | 0 |  
search.action.focusPreviousSearchResult | 0 |  
history.showNext | 0 |  
history.showPrevious | 0 |  
  |   |  
**Preferences** |   |  
workbench.action.openSettings | 0 |  
workbench.action.openWorkspaceSettings | 0 |  
workbench.action.openGlobalKeybindings | 0 |  
workbench.action.openSnippets | 0 |  
workbench.action.selectTheme | 0 |  
workbench.action.configureLocale | 0 |  
  |   |  
**Debug** |   |  
editor.debug.action.toggleBreakpoint | 1 |  
workbench.action.debug.start | 1 |  
workbench.action.debug.continue | 1 |  
workbench.action.debug.run | 1 |  
workbench.action.debug.pause | 1 |  
workbench.action.debug.stepInto | 1 |  
workbench.action.debug.stepOut | 1 |  
workbench.action.debug.stepOver | 1 |  
workbench.action.debug.stop | 1 |  
editor.debug.action.showDebugHover | Looks like it exists but can't find in keybinding widget? |  
  |   |  
**Tasks** |   |  
workbench.action.tasks.build | 0 |  
workbench.action.tasks.test | 0 |  
  |   |  
**Extensions** |   |  
workbench.extensions.action.installExtension | 0 |  
workbench.extensions.action.showInstalledExtensions | 0 |  
workbench.extensions.action.listOutdatedExtensions | 0 |  
workbench.extensions.action.showRecommendedExtensions | 0 |  
workbench.extensions.action.showPopularExtensions | 0 |  
workbench.extensions.action.updateAllExtensions | 0 |

**Comment by vince-fugnitto** [comment]
@JPinkney can you rebase your PR so that CI can pass?
It needs the following change https://github.com/theia-ide/theia/pull/5587

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 1]
```diff
@@ -90,9 +92,27 @@ export class CommandRegistryImpl implements CommandRegistryExt {
     executeCommand<T>(id: string, ...args: any[]): PromiseLike<T | undefined> {
         if (this.handlers.has(id)) {
             return this.executeLocalCommand(id, ...args);
-        } else {
+        } else if (KnownCommands.exclusions.hasOwnProperty(id)) {
```
Comment (marechal-p) [inline packages/plugin-ext/src/plugin/command-registry.ts]: If I understand correctly, only when an incoming `id` is in the `exclusion` map do we call `KnownCommands.map`, which because of this condition will branch in the first `exclusions[id]` check. Then my question is: When will it go in the `mappings[id]` branch?

File: packages/plugin-ext/src/plugin/types-impl.ts [hunk 2]
```diff
@@ -785,6 +785,23 @@ export class Location {
             this.range = new Range(rangeOrPosition, rangeOrPosition);
         }
     }
+
+    /*---------------------------------------------------------------------------------------------
+    *  Copyright (c) Microsoft Corporation. All rights reserved.
+    *  Licensed under the MIT License. See License.txt in the project root for license information.
+    *--------------------------------------------------------------------------------------------*/
+    // copied from https://github.com/microsoft/vscode/blob/master/src/vs/workbench/api/common/extHostTypes.ts
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/plugin/types-impl.ts]: We require a CQ for the copied code.
https://github.com/theia-ide/theia/wiki/Registering-CQs

File: packages/plugin-ext/src/common/object.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/object.ts]: 2019 please adjust all files

**Comment by akosyakov** [comment]
@JPinkney we should be able to parse keybindings from VS Code. It would be fine with me to register Monaco command twice as you suggested before, if there is no way to get rid of `monaco` prefix without breaking. It won't require maintaining mapping as well.

Very cool that you sorted out generic transformation between monaco and vscode APIs. Did you test it with code lenses? Could you provide how you test it?

File: packages/plugin-ext/src/plugin/types-impl.ts (hunk 2 above)
Comment (JPinkney) [reply packages/plugin-ext/src/plugin/types-impl.ts]: Do you have to be an official commiter in order to register a CQ? Following the document linked, I login and go to the project page but there isn't anywhere that says intellectual property or create a contribution questionnaire

**Comment by tsmaeder** [comment]
@vince-fugnitto @marechal-p @akosyakov what do we need to do to move this PR forward? File a CQ?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto @marechal-p @akosyakov what do we need to do to move this PR forward? File a CQ?

Yes, I believe a CQ should be registered for the copied code (only a committer can successfully register a CQ), the conflicts should be addressed, and perhaps document in the PR's description on how to test the PR? I think the only thing reviewers have done for the moment is to look at the source code.

**Comment by akosyakov** [comment]
@tsmaeder rebasing and providing `How to test` description that changes are verified independent, not by an author, then someone has to test it.

Also please file an issue that we should be able to parse VS Code keybindings, i.e. without `monaco` prefix. We can think how to handle it separately.

**Comment by tolusha** [comment]
@vince-fugnitto @akosyakov 
https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20661

> We have completed a high level triage of the attachment based on Parallel IP
> [1].  You may proceed to check the content into the Eclipse repository.  We
> will perform a comprehensive analysis at a later date based on IP work queue.

**Comment by akosyakov** [comment]
@tolusha that's cool, but PR needs rebase and explanation how to test from the user perspective.

As far as I understood fro a user there are 2 issues should be solved, now a command can be triggered from code lenses and custom keybindings can be used. Is it right? I assume @JPinkney used some extensions to check it. It would be good to share it to speed up review.

**Comment by akosyakov** [comment]
@JPinkney @tsmaeder Would you mind If I rebase it and write down test cases?

**Comment by tsmaeder** [comment]
@akosyakov @JPinkney is not shirking responsiblity, but got shanghaied into doing some OpenShift related work.

**Comment by akosyakov** [comment]
@JPinkney is it good to test?

**Comment by JPinkney** [comment]
@akosyakov Yes! I took out the VSCode bits I had in it because from what I understand it is only used for https://code.visualstudio.com/api/references/commands and none of them are implemented yet so there wasn't really a real test case against it

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 4]
```diff
@@ -102,9 +102,13 @@ export class CommandRegistryImpl implements CommandRegistryExt {
     executeCommand<T>(id: string, ...args: any[]): PromiseLike<T | undefined> {
         if (this.handlers.has(id)) {
             return this.executeLocalCommand(id, ...args);
-        } else {
+        } else if (KnownCommands.transformers.hasOwnProperty(id)) {
+            // Using the KnownCommand exclusions, convert the commands manually
             return KnownCommands.map(id, args, (mappedId: string, mappedArgs: any[] | undefined) =>
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/command-registry.ts]: Should not we change `map` function instead that it uses `KnownCommands.mappings[id]` if there is no `transformers`? 

The thing is that `map` is used to create commands for code lenses, quick fixes and so on. They also should call a properly mapped command.

File: packages/plugin-ext/src/common/known-commands.ts [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+// Here is a mapping of VSCode commands to monaco commands with their conversions
+export namespace KnownCommands {
+
+    /**
+     * Commands that you want to apply custom conversions to rather than pass through the automatic args converter.
+     * Would be useful in the case where theia provides some command and you need to provide custom conversions
+     */
+    // tslint:disable-next-line:no-any
+    export const transformers: { [id: string]: [string, (args: any[] | undefined) => any[] | undefined] } = {};
+    transformers['editor.action.showReferences'] = ['textEditor.commands.showReferences', createConversionFunction(
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/known-commands.ts]: Should not we remap it to `monaco.editor.action.showReferences` now and rely on generic conversion?

The issue here is that we test the PR with code lens references for java, but it won't use any mapping to monaco command.

File: packages/plugin-ext/src/common/known-commands.ts [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+// Here is a mapping of VSCode commands to monaco commands with their conversions
+export namespace KnownCommands {
+
+    /**
+     * Commands that you want to apply custom conversions to rather than pass through the automatic args converter.
+     * Would be useful in the case where theia provides some command and you need to provide custom conversions
+     */
+    // tslint:disable-next-line:no-any
+    export const transformers: { [id: string]: [string, (args: any[] | undefined) => any[] | undefined] } = {};
```
Comment (tsmaeder) [inline packages/plugin-ext/src/common/known-commands.ts]: I think the internal storage structure should not be part of the public interface: what we want to here is 2 things:
1. Transform an invocation of a know VS Code command into an invocation of a known Theia command (including parameter conversions)
2. Map the id of a VS Code command to the id of a Theia command. 

So we just need map(...) and mappedId(...)


We should always strive to have minimal interfaces and to centralize behaviour as much as possible. Exposing the internal structure creates coupling of the users of this class to the _implementation_ of this class. This breaks easily.

File: packages/plugin-ext/src/common/known-commands.ts [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        toArrayConversion(fromLocationToL))];
+
+    export function mapped(id: string): boolean {
+        return !!mappings[id];
+    }
+
+    /**
+     * Commands that have a name in theia that is different from their name in vscode
+     */
+    export const mappings: { [id: string]: string } = {};
```
Comment (tsmaeder) [inline packages/plugin-ext/src/common/known-commands.ts]: You really don't need a second array here: the "transformers" array already contains the mapping. The idea I have was this: if you just pass an empty array into "createConversionFunction", it will return the identity function (no conversion).  I think we can trust the JIT to optimize this case (if you are worried about performance) So 
transformers['id']= ['mappedId', createConversionFunction([])] should do the trick.

Also, memory is not an argument, as we are talking about < 100*8 bytes (100 references)

File: packages/plugin-ext/src/main/browser/keybindings/keybindings-contribution-handler.ts [hunk 8]
```diff
@@ -47,7 +47,16 @@ export class KeybindingsContributionPointHandler {
         if (!keybinding) {
             return undefined;
         }
-        const { command, when } = pluginKeybinding;
+        let { command, when } = pluginKeybinding;
```
Comment (tsmaeder) [inline packages/plugin-ext/src/main/browser/keybindings/keybindings-contribution-handler.ts]: This is the part the keeps us from doing the whole conversion on the plugin side. @akosyakov I think this is an argument for moving the parsing of contributions to the plugin side. I think that would also reduce the amount of data to be transferred over the wire (re: your discussion with @benoitf)
But of course, that is out of scope for this PR

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 4 above)
Comment (tsmaeder) [reply packages/plugin-ext/src/plugin/command-registry.ts]: I don't think there is a "should" in there: code action commands, etc. would not work with this code, right?

File: packages/plugin-ext/src/common/known-commands.ts [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+    mappings['editor.action.nextSelectionMatchFindAction'] = ['monaco.editor.action.nextSelectionMatchFindAction', createConversionFunction(...[])];
+    mappings['editor.action.previousSelectionMatchFindAction'] = ['monaco.editor.action.previousSelectionMatchFindAction', createConversionFunction(...[])];
+    mappings['editor.action.openLink'] = ['monaco.editor.action.openLink', createConversionFunction(...[])];
+    mappings['editor.action.quickFix'] = ['monaco.editor.action.quickFix', createConversionFunction(...[])];
+    mappings['editor.action.refactor'] = ['monaco.editor.action.refactor', createConversionFunction(...[])];
+    mappings['editor.action.sourceAction'] = ['monaco.editor.action.sourceAction', createConversionFunction(...[])];
+    mappings['editor.action.organizeImports'] = ['monaco.editor.action.organizeImports', createConversionFunction(...[])];
+    mappings['editor.action.autoFix'] = ['monaco.editor.action.autoFix', createConversionFunction(...[])];
+    mappings['editor.action.fixAll'] = ['monaco.editor.action.fixAll', createConversionFunction(...[])];
+    mappings['editor.action.rename'] = ['monaco.editor.action.rename', createConversionFunction(...[])];
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/known-commands.ts]: What about conversions between 0-based positions/locations and 1-based? This command will produce bogus results without conversions.

I thought this PR used to have generic conversion similar to what VS Code does: https://github.com/Microsoft/vscode/blob/70160d0e580d0588e37c93fe8724014d76b00939/src/vs/workbench/api/node/extHostCommands.ts#L110-L138

One can test it with extract refactorings for typescript vscode extension.

File: packages/plugin-ext/src/main/browser/keybindings/keybindings-contribution-handler.ts (hunk 8 above)
Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/keybindings/keybindings-contribution-handler.ts]: Ideally the plugin system should register needed vscode commands and monaco commands without `monaco.` prefix in Theia command registry as well, otherwise we are not compatible with keybinding preferences for VS Code defined by users. If we have all vscode and monaco commands in our registry then mapping on the plugin side is not necessary. We need only apply generic conversion of positions/locations.

For performance one has to measure first whether there is an issue in this case.

**Comment by akosyakov** [comment]
After testing a bit. I also wonder whether this code is necessary:
https://github.com/JPinkney/theia/blob/16da2d93f199670274b0e1b68e5a3ae6e7cb4069/packages/plugin-ext/src/plugin/command-registry.ts#L168-L170

Or it should be `if (KnownCommands.mapped((command.command || command.id) as string)) {`

It does not break, since the safe command does the mapped check properly eventually, but it does not seem ever go in if-branch as for now.

File: packages/plugin-ext/src/common/known-commands.ts (hunk 9 above)
Comment (tsmaeder) [reply packages/plugin-ext/src/common/known-commands.ts]: No, the idea was always to explicitly convert. That way, it also works for stuff registered by clients that are not targeted to monaco.
> // automagically convert some argument types

automagically: brr.....

File: packages/plugin-ext/src/common/known-commands.ts [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    mappings['editor.action.insertCursorBelow'] = ['monaco.editor.action.insertCursorBelow', createConversionFunction(...[])];
+    mappings['editor.action.insertCursorAtEndOfEachLineSelected'] = ['monaco.editor.action.insertCursorAtEndOfEachLineSelected', createConversionFunction(...[])];
+    mappings['editor.action.addSelectionToNextFindMatch'] = ['monaco.editor.action.addSelectionToNextFindMatch', createConversionFunction(...[])];
+    mappings['editor.action.addSelectionToPreviousFindMatch'] = ['monaco.editor.action.addSelectionToPreviousFindMatch', createConversionFunction(...[])];
+    mappings['editor.action.moveSelectionToNextFindMatch'] = ['monaco.editor.action.moveSelectionToNextFindMatch', createConversionFunction(...[])];
+    mappings['editor.action.moveSelectionToPreviousFindMatch'] = ['monaco.editor.action.moveSelectionToPreviousFindMatch', createConversionFunction(...[])];
+    mappings['editor.action.selectHighlights'] = ['monaco.editor.action.selectHighlights', createConversionFunction(...[])];
+    mappings['editor.action.changeAll'] = ['monaco.editor.action.changeAll', createConversionFunction(...[])];
```
Comment (tsmaeder) [inline packages/plugin-ext/src/common/known-commands.ts]: Does the compiler complain about "createConversionFunction()", also, I think we can use a constant IDENTITY= createConversionFunction()

File: packages/plugin-ext/src/common/known-commands.ts (hunk 9 above)
Comment (akosyakov) [reply packages/plugin-ext/src/common/known-commands.ts]: It is just we need to review then all mapped commands and make sure that conversions are proper. Should it be a part of this PR?

Comment (tsmaeder) [reply packages/plugin-ext/src/common/known-commands.ts]: Well, "automagic" conversion needs to be tested, as well. I believe all the monaco action need the same conversions (0- vs 1-based). Problem is: we're not using the monaco location conventions everywhere in Theia, so I don't think we can generally convert parameters.
What about a plugin registering a command that is invoked by another plugin (with positional parameters). All those considerations led me to decide against the automatic conversion. I know it's more code, but safer, IMO.

Comment (akosyakov) [reply packages/plugin-ext/src/common/known-commands.ts]: > What about a plugin registering a command that is invoked by another plugin (with positional parameters).

We could apply it only if a command is mapped to monaco command, i.e. we define a single monaco command conversion function and pass it as a second argument: `['monaco.editor.action.rename', monacoConversion]`. We should also not forget to convert a result.

**Comment by svenefftinge** [comment]
We have other commands registered here : https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts#L166

I don't care too much whether these commands are in the vscode layer or not, but it would be great to have them registered in one place.

**Comment by JPinkney** [comment]
@akosyakov @svenefftinge I'm attempting to register the monaco commands without the prefix [here](https://github.com/eclipse-theia/theia/blob/master/packages/monaco/src/browser/monaco-command-registry.ts#L51) and after rebuilding I'm getting `Maximum call stack size exceeded` when I open a file in theia and click backspace. I thought this was working before, was there anything that changed recently in regards to this area that might be causing this?

**Comment by benoitf** [comment]
https://github.com/eclipse-theia/theia/commit/821b9dc27653ea971f77e2b1f3771b7a4e44b1e0

**Comment by akosyakov** [comment]
@JPinkney is it good to have a look again?

**Comment by JPinkney** [comment]
@akosyakov Yes

**Comment by akosyakov** [comment]
I'm very busy this week. @eclipse-theia/eclipse-theia Could someone help testing it?
-  @JPinkney defined some tests in `How to test` section of the description.
- I wanted to test also typescript VS Code extension:
  - typesctipt VS Code extension for testing: [ts_plugins.zip](https://github.com/theia-ide/theia/files/3513239/ts_plugins.zip)
  - uninstall native typescript extension
  - With this PR refactorings like extract method, variable and so on should work. Pay attention is there any issue with positions, i.e. they should be bogusly shifted by 1 character.

**Comment by elaihau** [comment]
I am trying to test this pull request with the typescript VS Code extension. 

Could Anton @akosyakov or someone kindly provide me some guidance on how to uninstall native Theia typescript extension ? 

Thank you in advance !

**Comment by kittaakos** [comment]
> how to uninstall native Theia typescript extension

You can remove the `@theia/typescript` dependency from the example app's `package.json`.

**Comment by akosyakov** [comment]
@JPinkney Could you remind pleas why cannot we drop `monaco.` prefix and register Monaco command as is once?

File: packages/plugin-ext/src/common/objects.ts [hunk 11]
```diff
@@ -0,0 +1,233 @@
+// tslint:disable
+// copied from https://github.com/microsoft/vscode/blob/1.37.0/src/vs/base/common/objects.ts
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import { isUndefinedOrNull, isArray, isObject } from './types';
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/objects.ts]: this file has a lot of unused code, please remove as much as possible

also check whether we already have required methods in our dependencies or in core, i.e. deepClone and deepFreeze can be found in core

File: packages/plugin-ext/src/common/types.ts [hunk 12]
```diff
@@ -75,3 +75,101 @@ export function es5ClassCompat<T extends Function>(target: T): T {
     return _ as unknown as T;
 }
 // tslint:enable:no-any
+
+// copied from https://github.com/microsoft/vscode/blob/1.37.0/src/vs/base/common/types.ts
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+const _typeof = {
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/types.ts]: please get rid of unused copied code

File: packages/plugin-ext/src/plugin/types-impl.ts [hunk 13]
```diff
@@ -785,6 +785,22 @@ export class Location {
             this.range = new Range(rangeOrPosition, rangeOrPosition);
         }
     }
+
+    // copied from https://github.com/microsoft/vscode/blob/1.37.0/src/vs/workbench/api/common/extHostTypes.ts
+    /*---------------------------------------------------------------------------------------------
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/types-impl.ts]: copyrings should be declared at  the top of a file, please review other places

**Comment by JPinkney** [comment]
@akosyakov In the PR it is registering without monaco prefix as well: https://github.com/eclipse-theia/theia/pull/5590/files#diff-3e78733f92a6f6b63b219e5fec2097f6R55 and then it hides anything with the monaco prefix in the keybinding view. From what I understood we had to keep the monaco prefix because other extenders of theia might be using it?

**Comment by akosyakov** [comment]
> From what I understood we had to keep the monaco prefix because other extenders of theia might be using it?

We can ask extenders to drop monaco prefix. It was more to avoid changing core and monaco extensions only apply changes to the plugin since. But since we cannot avoid it, i wonder can we just drop monaco prefix? Otherwise we introduce for example an implicit dependency from core to monaco by checking kebindings.

There are excluded commands which should not be registered. I think we should keep them but rather than that rest of monaco commands can be registered without a prefix.

**Comment by akosyakov** [comment]
@JPinkney let me know when it is possible to test again. I see that tests are failing. Please also search through the repo for `monaco.` prefixes. We should drop them.

**Comment by akosyakov** [comment]
@JPinkney is it good now for testing? there seem to be some conflicting files

**Comment by JPinkney** [comment]
@akosyakov Just fixed the conflict, should be good to test now

**Comment by akosyakov** [comment]
Oh, I suspect the issue is that no keybinding matches in our registry, so then the default Monaco keyboard handler is called. 🙈 We should replace Monaco keybinding service with the null implementation.

**Comment by akosyakov** [comment]
Adding a record to CHANGELOG with a breaking change that `monaco.` prefix was dropped would be helpful.
